Solution redux

StrongerProgrammer7 opened the pull request:
Added React-Redux
Added
- Store
- Models: User: userReducer and Controls
- Enums: ActionTypes
- Interfaces: IAction
Deleted all Context

StrongerProgrammer7 committed:
Create Modal window for settings:
create useStates for inputs
Create Context for store data user
Create Button

StrongerProgrammer7 committed:
Created useState for record user's data
Refactor: create MyInput
create debounce for inputs
Change interface IDataUser:
create function: useState

StrongerProgrammer7 committed:
Create method saveChanges
create interface for input's data
delete available null for show and close modal

StrongerProgrammer7 committed:
Added octokit
Create helper for work localstorage
add method setDataContextFromLocalStorage
Refactor MyPopup  (create file utils.ts)
create interface User
Made IDatauser extends IUser

StrongerProgrammer7 committed:
Made read from localStorage
Save to Context from localStorage
made save to localStorage
Add show spinner read data

StrongerProgrammer7 committed:
Solution:
change github token
Added realization get contributors
Change myButton
Created utils for myPopup
Created interface IContributor
Created interface ISetInputs

StrongerProgrammer7 committed:
Added Redux: React-Redux
Added interface Action
Added enums ActionTypes
Added store
Added model: user (
initialState,
userReducer,
Controls)
Change all function for work with store

StrongerProgrammer7 committed:
Added method for work with githubAPI Get

StrongerProgrammer7 committed:
Added Prettier and ESLint
Accept format

StrongerProgrammer7 committed:
Deleted Context

aartem1 commented on src/components/UI/Buttons/MyButton.tsx:
```diff
@@ -0,0 +1,28 @@
+  callback
+}: {
+  title: string;
+  callback?: () => void;
```
Кнопка без callback, кажется не имеет смысла. Возможно, какие-то частные случаи могут быть, но кажется можно сделать required.

StrongerProgrammer7 replied:
Fix

StrongerProgrammer7 committed:
Fix some probelm:
MyButton:
change call function on the Click
delete duplicata define type
MyInput fix with accept trim
MyPopup: deleted useEffect
Deleted eslint so as there is problem

StrongerProgrammer7 committed:
Fix: delete useState:
create StandartInput using ref
MyPopup:
Deleted useState
using useRef
Made button submit
Made onSubmit form
Action Types:
Added types for found reviewer
IAction added interface for reviewer
IDataUser added IReviewer
Added IReducers for including combine reducers
Added combine reducers
Create reviewerReducer and controls

StrongerProgrammer7 committed:
Added reducer for Loadings
Move and decomposition IAction

StrongerProgrammer7 committed:
Deleted not need interface IReviewer
Added redux-thunk to package.json
Added simple middleware

StrongerProgrammer7 committed:
Added middleware

StrongerProgrammer7 committed:
Accept Prettier

StrongerProgrammer7 committed:
Not working version for reviewer

StrongerProgrammer7 committed:
FIx problem: work async with middleware

StrongerProgrammer7 committed:
🐞 fix(Fix middleware): Fix and refactor

Created useTypedSelector ;
Created useAppDispatch;
deleted IReducers and added typeof dispatch and store ;

aartem1 commented on src/App.tsx:
```diff
@@ -1,25 +1,70 @@
+  const reviewer = useTypedSelector((state) => state.reviewerReducer);
+  const loadings = useTypedSelector((state) => state.loadingsReducer);
+  const dispatch = useAppDispatch();
+  const generateReviewer = useRef<null | HTMLImageElement>(null);
```
Хранить такое на уровне компонента не совсем правильно
Если ты хочешь реализовать  логику имитации перебора ревьюеров, то это тоже должно быть в сторе, так как иначе получается размазывание логики хранения состояния

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/components/UI/Buttons/MyButton.tsx:
```diff
@@ -0,0 +1,15 @@
+import React, { FC } from 'react';
+import { Button } from 'react-bootstrap';
+
+const MyButton: FC<{ title: string; callback?: () => void }> = ({ title, callback }) => {
```
Название "MyButton" ничего не говорит о том, что это за кнопка
Лучше тогда это сделать просто "Button" если это абстрактная базовая кнопка

aartem1 replied:
Чтобы не было конфликтов с бутстрапом можно сделать
`import { Button as BootstrapButton } from 'react-bootstrap';`

StrongerProgrammer7 replied:
Предполагается что для этой кнопки будут свои стили и возможно функционал, это позволит легко это все сделать отсюда и MyButton.  Так как проект простой , сама кнопка простая :)

aartem1 commented on src/components/UI/Inputs/MyInput.tsx:
```diff
@@ -0,0 +1,54 @@
+  };
+};
+
+export const MyInput = ({
```
этот компонент кажется вообще не используется

StrongerProgrammer7 replied:
Этот компонент от  ДЗ React , deleted

aartem1 commented on src/components/UI/Inputs/StandartInput.tsx:
```diff
@@ -0,0 +1,30 @@
+}: {
+  titleLabel: string;
+  placeholder: string;
+  _defaultValue: string;
```
зачем здесь проп с нижним подчеркиванием?

StrongerProgrammer7 replied:
чтобы не было пересечения с defaultValue и было удобнее получить доступ

aartem1 replied:
Каких пересечений? у тебя ниже `defaultValue={_defaultValue}`. Кажется что нет смысла тут делать разные названия

И _ в именах пропов не совсем стандартный паттерн, во всяком случае для ref и defaultValue

StrongerProgrammer7 replied:
Fix

aartem1 replied:
Пересекаться может только ref, обычно в таком случае оборачивают в forwardRef, либо можно назвать твой проп innerRef
defaultValue вроде бы не имеет пересечений ни с чем стандартным, его можно сделать просто вот так

`export const StandartInput = ({
  titleLabel,
  placeholder,
  defaultValue,
  _ref: ref,
  type = 'text'
}`

StrongerProgrammer7 replied:
Using forwardRef

aartem1 commented on src/components/UI/Popup/MyPopup.tsx:
```diff
@@ -0,0 +1,72 @@
+import { IWorkModal } from '../../../store/interfaces/IWorkModal';
+import { StandartInput } from '../Inputs/StandartInput';
+import { saveChanges } from './utils';
+const MyPopup: FC<IWorkModal> = ({ show, handleClose }: IWorkModal) => {
```
Тебе нужно решить, это просто абстрактный компонент попапа, грубо говорят элемент дизайн системы, тогда в нем не должно быть подключение стора напрямую, а должны прокидываться пропы

Либо это компонент реализующий бизнес логику, но тогда у него должно быть более понятное название, а не MyPopup

StrongerProgrammer7 replied:
Согласен, переименовал

aartem1 commented on src/components/UI/Popup/utils.ts:
```diff
@@ -0,0 +1,23 @@
+  data: IInputs,
+  handleClose: (() => void) | undefined
+): void => {
+  dispatch(UserControls.changeLogin(data.loginInput) as IActionUser);
```
Кажется нет особого смысла тут делить эти данные на три диспатча.
Они всегда используются вместе, можно объединить в один action, код станет проще.

StrongerProgrammer7 replied:
Добавил в UserControls setAllData
Теперь используется один dispatch

aartem1 commented on src/models/loading.ts:
```diff
@@ -0,0 +1,29 @@
+  loadReadyShowReviewer: false
+};
+
+export function loadingsReducer(state = initialState, action: IActionLoadings) {
```
обычно признак загрузки данных кладут не в отдельный редьюсер, а рядом с теми данные, которые грузятся, в одном редьюсере

StrongerProgrammer7 replied:
Да, но здесь я подразумеваю общий "Загрузчик" ,
В данном случае обычный "Загрузчик" для большинства данных ,
А вот для ревьювера , тут уже для генерации ревьевера, но я подумал, что сложу в одни reducer все "Загрузчики" и буду от туда их доставать

aartem1 replied:
Я бы все-таки предложил использовать подход, когда признак загрузки лежит рядом с данными и отказаться от loadingReducer

Это выглядит не стандартно. Обычно редьюсер это какая-то сущность и все данные относящиеся к этой сущности хранятся в нем. Тогда сразу понятно какой редьюсер смотреть, если нужно работать с этими данными

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/models/reviewer.ts:
```diff
@@ -0,0 +1,30 @@
+}
+
+export const reviewerControls = {
+  changeLoginReviewer: (value: string) => ({
```
Зачем здесь два экшена?
Ревьюер это единая сущность, его можно диспатчить в стор за один экшен

StrongerProgrammer7 replied:
С одной стороны да, а с другой, вдруг Гит решит добавить фичу несколько аватаров ??? ))) 
А у нас тут есть action поменять аватар )) 
Fix - now one Action

aartem1 commented on src/store/enums/ActionTypes.ts:
```diff
@@ -0,0 +1,9 @@
+  BLACKLIST = 'BLACKLIST',
+  REVIEWER = 'REVIEWER',
+  AVATAR = 'AVATAR',
+  BASELOADING = 'BASELOADING',
```
`BASE_LOADING` будет лучше читаться, аналогично другие константы тоже

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/store/interfaces/IDataUser.ts:
```diff
@@ -0,0 +1,9 @@
+  blacklist: Array<string>;
+}
+export interface IContributor {
+  avatar_url: string;
```
лучше везде использовать один стиль. Если в других местах камел кейс, то должно быть `avatarUrl`

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/store/interfaces/IWorkModal.ts:
```diff
@@ -0,0 +1,5 @@
+export interface IWorkModal {
+  show: boolean;
+  handleShow?: () => void;
```
есть тип `VoidFunction`

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/utils/helper.ts:
```diff
@@ -0,0 +1,140 @@
+
+  return getRandomNumber(0, contributors.length - 1);
+}
+function stopInterval(interval: NodeJS.Timer) {
```
в этой обертке кажется нет смысла

StrongerProgrammer7 replied:
Fix

StrongerProgrammer7 committed:
Fix MyButton : make required callback

StrongerProgrammer7 committed:
Deleted MyInput

StrongerProgrammer7 committed:
Rename MyPopup-> PopupSettings

StrongerProgrammer7 committed:
Added:
interface:IActionTotalData
enums:TOTAL_DATA_USER
controls: setAllData
change dispatch in saveChanges

StrongerProgrammer7 committed:
Made VoidFunction

StrongerProgrammer7 committed:
Fix:
Change Action Reviewer - one Action

StrongerProgrammer7 committed:
Deleted not need method stopInterval

StrongerProgrammer7 committed:
Added right type:
AsyncDispatch

StrongerProgrammer7 committed:
🦄 refactor: Generate Reviewer

Deleted useRef and added generateReviewerImgSrc to store;
Change method showAndChooseReviewer

aartem1 commented on src/utils/helper.ts:
```diff
@@ -0,0 +1,134 @@
+  ) =>
+  async (dispatch: Dispatch<IActionReviewer | IActionLoadings>, getState: () => RootState) => {
+    const user = getState().userReducer;
+    dispatch(loadingsControls.changeBaseLoad(true) as IActionLoadings);
```
Лоадинг все-таки должен быть привязан к модели данных
Не нужно разделять лоадинг и данные на два редьюсера

Потому что иначего получается странное флоу, у тебя можно выполнить экшен смены ревьюера, но без показа загрузки

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/App.tsx:
```diff
@@ -1,25 +1,69 @@
+
+  useEffect(() => {
+    setDataFromLocalStorage(dispatch, 'userdata');
+    dispatch(loadingsControls.changeBaseLoad(false) as IActionLoadings);
```
зачем это нужно?
У тебя по дефолту оно false по идее

StrongerProgrammer7 replied:
Так как загрузчик общий, то по дефолту он true 
как бы загрузка данных с localStorage , как закончили загрузку, убрали лоадинг и отобразили страницу для взаимодействия

aartem1 commented on src/App.tsx:
```diff
@@ -1,25 +1,69 @@
-      </header>
-    </div>
+    <>
+      {loadings.baseLoad === true ? (
```
можно убрать `=== true`

StrongerProgrammer7 replied:
Не думаю, что это сильно влияет, но ок fix

aartem1 commented on src/components/UI/Popup/PopupSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+import 'bootstrap/dist/css/bootstrap.min.css';
```
Скорее всего, ты что-то не правильно сделал при подключении этой библиотеки, если есть необходимость импортировать это в каждый компонент. Обычно такое импортируется в рутовый компонент или в глобальные стили

StrongerProgrammer7 replied:
Даже не заметил как появилось, оно вообще не нужно, удалил

aartem1 commented on src/components/UI/Popup/PopupSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+          if (!loginInput || !repoInput || !blacklistInput) return;
+          if (!loginInput.current || !repoInput.current || !blacklistInput.current) return;
+
+          saveChanges(
```
спорный паттерн переносить такие простые хелперы в отдельные файлы. Знаю, что в некоторых проектах принято хранить в tsx только верстку, а в utils/helpers уносить вообще любые функции. Но тут как будто бы это избыточно и ухудшает чтение кода. Потому что приходится лезть в другой файл смотреть, что за функция saveChanges, а в ней по сути три строчки, которые можно было и тут написать и используется эта функция в одном только месте.

 Но это вкусовщина, я не настаиваю на изменении.

aartem1 commented on src/components/UI/Popup/utils.ts:
```diff
@@ -0,0 +1,21 @@
+  data: IInputs,
+  handleClose: (() => void) | undefined
+): void => {
+  dispatch(UserControls.setAllData({login:data.loginInput,repo:data.repoInput,blacklist:dataFromStringToArray(data.blacklistInput)}) as IActionUser)
```
А тебе тут все еще нужно приведение типов явное разве?

StrongerProgrammer7 replied:
Да, у  меня TS ругается без этого

aartem1 replied:
Потому что тебе надо поправить типизацию

Во первых у тебя не хватает типа вот тут

```
type IActionUser =
  | IActionLogin
  | IActionREPO
  | IActionBlackList
  | IActionTotalData
  | IActionLoad
  | { type: ActionTypes.TOTAL_DATA_USER; payload: IUser };
```
Последняя строчка
Фикшу максимально просто, придумай названия типов сам

Во вторых тебе нужно типизировать экшен примерно вот так

```
export const setAllData = (
  value: IUser
): { type: ActionTypes.TOTAL_DATA_USER; payload: IUser } => ({
  type: ActionTypes.TOTAL_DATA_USER,
  payload: value
});
```

Потому что без этого указания выглядит так что твой action creator может вернуть любой ActionTypes

StrongerProgrammer7 replied:
Тот тип который ты написал это :IActionTotalData )) 
Получается я просто везде должен указать возвращение функции

aartem1 replied:
Да у экшен креэйтора при типизации нужно указывать возвращаемый тип явно либо использовать какие-то надстройки (например, редакс тулкит :) )

StrongerProgrammer7 replied:
Ха, мощный рофл, тулкит нельзя ведь ))) 
Fix

aartem1 commented on src/components/UI/Popup/utils.ts:
```diff
@@ -0,0 +1,21 @@
+    'userdata',
+    data.loginInput + ';' + data.repoInput + ';' + data.blacklistInput
+  );
+  if (handleClose) handleClose();
```
можно это писать проще как `handleClose?.()`

StrongerProgrammer7 replied:
Fix

aartem1 commented on src/hooks/customHooks/useGetComplexObject.jsx:
```diff
@@ -0,0 +1,12 @@
+import { useContext } from 'react';
+import { Context } from '../../App';
+
+const useGetComplexObject = () => {
```
А это зачем? По ПРу не вижу использования

StrongerProgrammer7 replied:
Со старого проекта, удалил

aartem1 commented on src/store/enums/ActionTypes.ts:
```diff
@@ -0,0 +1,10 @@
+  REVIEWER = 'REVIEWER',
+  GENERATE_REVIEWER_IMGS = 'GENERATE_REVIEWER',
+  BASE_LOADING = 'BASELOADING',
+  LOAD_SHOW_REVIEWER = 'LOADREADYSHOWREVIEWER'
```
Подчеркивания

StrongerProgrammer7 replied:
Fix

StrongerProgrammer7 committed:
Deleted _

StrongerProgrammer7 committed:
Deleted not need hooks for Context

StrongerProgrammer7 committed:
Deleted import bootstrap.css
Accept ?.

StrongerProgrammer7 committed:
Added _ for ActionTypes

StrongerProgrammer7 committed:
Deleted loading reducer
Move loadShowReviewer to ReviewerReducer
Move baseLoad to UserReducer

StrongerProgrammer7 committed:
Fix small problem with boolean loading

StrongerProgrammer7 committed:
Using forwardRef
Deleted _

StrongerProgrammer7 committed:
Fix problem with forwardRef
Made return type for all action creator

StrongerProgrammer7 committed:
Deleted as type: as unkown as IAction

StrongerProgrammer7 committed:
Settings eslint prettier
